File: sdk/python/urlshortener/services/analytics.py
"""Analytics service for URL Shortener API."""

from typing import List, Optional, Dict, Any
from datetime import datetime

from ..http_client import HTTPClient, AsyncHTTPClient
from ..types.analytics import (
    Click,
    URLAnalytics,
    DashboardStats,
    GeoStats,
    DeviceStats,
    ReferrerStats,
    AnalyticsFilter,
    AnalyticsExportRequest,
    RealTimeStats,
    ComparisonStats,
    ClickHeatmap,
    AnalyticsInsights,
    CountryStat,
    CityStat,
    DeviceStat,
    BrowserStat,
    OStat,
    ReferrerStat,
    TimelineData
)
from ..types.common import RequestConfig


class AnalyticsService:
    """Synchronous analytics service."""

    def __init__(self, http_client: HTTPClient):
        """Initialize analytics service.
        
        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def get_dashboard_stats(
        self,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> DashboardStats:
        """Get dashboard overview statistics.
        
        Args:
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            DashboardStats: Dashboard statistics
        """
        params = {}
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/dashboard',
            params=params,
            config=config
        )
        
        return DashboardStats(**response['data'])

    def get_url_analytics(
        self,
        url_id: int,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        config: Optional[RequestConfig] = None
    ) -> URLAnalytics:
        """Get comprehensive analytics for a specific URL.
        
        Args:
            url_id: URL ID
            start_date: Start date for analytics
            end_date: End date for analytics
            config: Request configuration
            
        Returns:
            URLAnalytics: Comprehensive URL analytics
        """
        params = {}
        if start_date:
            params['start_date'] = start_date.isoformat()
        if end_date:
            params['end_date'] = end_date.isoformat()
            
        response = self.http_client.get(
            f'/analytics/urls/{url_id}',
            params=params,
            config=config
        )
        
        return URLAnalytics(**response['data'])

    def get_clicks(
        self,
        url_id: Optional[int] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        limit: int = 100,
        offset: int = 0,
        config: Optional[RequestConfig] = None
    ) -> List[Click]:
        """Get individual click records.
        
        Args:
            url_id: Filter by URL ID
            start_date: Start date filter
            end_date: End date filter
            limit: Maximum number of records
            offset: Number of records to skip
            config: Request configuration
            
        Returns:
            List[Click]: List of click records
        """
        params = {
            'limit': limit,
            'offset': offset
        }
        
        if url_id:
            params['url_id'] = url_id
        if start_date:
            params['start_date'] = start_date.isoformat()
        if end_date:
            params['end_date'] = end_date.isoformat()
            
        response = self.http_client.get(
            '/analytics/clicks',
            params=params,
            config=config
        )
        
        return [Click(**item) for item in response['data']]

    def get_geographic_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> GeoStats:
        """Get geographic statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            GeoStats: Geographic statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/geo',
            params=params,
            config=config
        )
        
        return GeoStats(**response['data'])

    def get_device_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> DeviceStats:
        """Get device and browser statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            DeviceStats: Device statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/devices',
            params=params,
            config=config
        )
        
        return DeviceStats(**response['data'])

    def get_referrer_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> ReferrerStats:
        """Get referrer statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            ReferrerStats: Referrer statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/referrers',
            params=params,
            config=config
        )
        
        return ReferrerStats(**response['data'])

    def get_timeline_data(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        granularity: str = 'day',
        config: Optional[RequestConfig] = None
    ) -> List[TimelineData]:
        """Get timeline data for clicks.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            granularity: Data granularity (hour, day, week, month)
            config: Request configuration
            
        Returns:
            List[TimelineData]: Timeline data points
        """
        params = {
            'granularity': granularity
        }
        
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/timeline',
            params=params,
            config=config
        )
        
        return [TimelineData(**item) for item in response['data']]

    def get_real_time_stats(
        self,
        config: Optional[RequestConfig] = None
    ) -> RealTimeStats:
        """Get real-time analytics statistics.
        
        Args:
            config: Request configuration
            
        Returns:
            RealTimeStats: Real-time statistics
        """
        response = self.http_client.get('/analytics/realtime', config=config)
        return RealTimeStats(**response['data'])

    def get_comparison_stats(
        self,
        url_id: int,
        current_start: datetime,
        current_end: datetime,
        previous_start: datetime,
        previous_end: datetime,
        config: Optional[RequestConfig] = None
    ) -> ComparisonStats:
        """Compare analytics between two time periods.
        
        Args:
            url_id: URL ID
            current_start: Current period start date
            current_end: Current period end date
            previous_start: Previous period start date
            previous_end: Previous period end date
            config: Request configuration
            
        Returns:
            ComparisonStats: Comparison statistics
        """
        params = {
            'current_start': current_start.isoformat(),
            'current_end': current_end.isoformat(),
            'previous_start': previous_start.isoformat(),
            'previous_end': previous_end.isoformat()
        }
        
        response = self.http_client.get(
            f'/analytics/urls/{url_id}/compare',
            params=params,
            config=config
        )
        
        return ComparisonStats(**response['data'])

    def get_click_heatmap(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> ClickHeatmap:
        """Get click heatmap data.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            ClickHeatmap: Heatmap data
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/heatmap',
            params=params,
            config=config
        )
        
        return ClickHeatmap(**response['data'])

    def get_insights(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> AnalyticsInsights:
        """Get analytics insights and recommendations.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            AnalyticsInsights: Analytics insights
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/insights',
            params=params,
            config=config
        )
        
        return AnalyticsInsights(**response['data'])

    def export_analytics(
        self,
        url_ids: Optional[List[int]] = None,
        format: str = 'csv',
        period: Optional[str] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        include_details: bool = False,
        filters: Optional[AnalyticsFilter] = None,
        config: Optional[RequestConfig] = None
    ) -> Dict[str, Any]:
        """Export analytics data.
        
        Args:
            url_ids: Specific URL IDs to export
            format: Export format (csv, json, xlsx)
            period: Time period (24h, 7d, 30d, etc.)
            start_date: Start date for export
            end_date: End date for export
            include_details: Include detailed click data
            filters: Additional filters
            config: Request configuration
            
        Returns:
            Dict[str, Any]: Export information
        """
        request = AnalyticsExportRequest(
            url_ids=url_ids,
            format=format,
            period=period,
            start_date=start_date,
            end_date=end_date,
            include_details=include_details,
            filters=filters
        )
        
        response = self.http_client.post(
            '/analytics/export',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return response['data']

    def get_top_countries(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        limit: int = 10,
        config: Optional[RequestConfig] = None
    ) -> List[CountryStat]:
        """Get top countries by clicks.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            limit: Number of results
            config: Request configuration
            
        Returns:
            List[CountryStat]: Top countries
        """
        params = {
            'limit': limit
        }
        
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/top/countries',
            params=params,
            config=config
        )
        
        return [CountryStat(**item) for item in response['data']]

    def get_top_cities(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        limit: int = 10,
        config: Optional[RequestConfig] = None
    ) -> List[CityStat]:
        """Get top cities by clicks.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            limit: Number of results
            config: Request configuration
            
        Returns:
            List[CityStat]: Top cities
        """
        params = {
            'limit': limit
        }
        
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/top/cities',
            params=params,
            config=config
        )
        
        return [CityStat(**item) for item in response['data']]

    def get_top_browsers(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        limit: int = 10,
        config: Optional[RequestConfig] = None
    ) -> List[BrowserStat]:
        """Get top browsers by clicks.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            limit: Number of results
            config: Request configuration
            
        Returns:
            List[BrowserStat]: Top browsers
        """
        params = {
            'limit': limit
        }
        
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = self.http_client.get(
            '/analytics/top/browsers',
            params=params,
            config=config
        )
        
        return [BrowserStat(**item) for item in response['data']]


class AsyncAnalyticsService:
    """Asynchronous analytics service."""

    def __init__(self, http_client: AsyncHTTPClient):
        """Initialize async analytics service.
        
        Args:
            http_client: Async HTTP client instance
        """
        self.http_client = http_client

    async def get_dashboard_stats(
        self,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> DashboardStats:
        """Get dashboard overview statistics.
        
        Args:
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            DashboardStats: Dashboard statistics
        """
        params = {}
        if period:
            params['period'] = period
            
        response = await self.http_client.get(
            '/analytics/dashboard',
            params=params,
            config=config
        )
        
        return DashboardStats(**response['data'])

    async def get_url_analytics(
        self,
        url_id: int,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        config: Optional[RequestConfig] = None
    ) -> URLAnalytics:
        """Get comprehensive analytics for a specific URL.
        
        Args:
            url_id: URL ID
            start_date: Start date for analytics
            end_date: End date for analytics
            config: Request configuration
            
        Returns:
            URLAnalytics: Comprehensive URL analytics
        """
        params = {}
        if start_date:
            params['start_date'] = start_date.isoformat()
        if end_date:
            params['end_date'] = end_date.isoformat()
            
        response = await self.http_client.get(
            f'/analytics/urls/{url_id}',
            params=params,
            config=config
        )
        
        return URLAnalytics(**response['data'])

    async def get_clicks(
        self,
        url_id: Optional[int] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        limit: int = 100,
        offset: int = 0,
        config: Optional[RequestConfig] = None
    ) -> List[Click]:
        """Get individual click records.
        
        Args:
            url_id: Filter by URL ID
            start_date: Start date filter
            end_date: End date filter
            limit: Maximum number of records
            offset: Number of records to skip
            config: Request configuration
            
        Returns:
            List[Click]: List of click records
        """
        params = {
            'limit': limit,
            'offset': offset
        }
        
        if url_id:
            params['url_id'] = url_id
        if start_date:
            params['start_date'] = start_date.isoformat()
        if end_date:
            params['end_date'] = end_date.isoformat()
            
        response = await self.http_client.get(
            '/analytics/clicks',
            params=params,
            config=config
        )
        
        return [Click(**item) for item in response['data']]

    async def get_geographic_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> GeoStats:
        """Get geographic statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            GeoStats: Geographic statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = await self.http_client.get(
            '/analytics/geo',
            params=params,
            config=config
        )
        
        return GeoStats(**response['data'])

    async def get_device_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> DeviceStats:
        """Get device and browser statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            DeviceStats: Device statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = await self.http_client.get(
            '/analytics/devices',
            params=params,
            config=config
        )
        
        return DeviceStats(**response['data'])

    async def get_referrer_stats(
        self,
        url_id: Optional[int] = None,
        period: Optional[str] = None,
        config: Optional[RequestConfig] = None
    ) -> ReferrerStats:
        """Get referrer statistics.
        
        Args:
            url_id: Filter by URL ID
            period: Time period (24h, 7d, 30d, etc.)
            config: Request configuration
            
        Returns:
            ReferrerStats: Referrer statistics
        """
        params = {}
        if url_id:
            params['url_id'] = url_id
        if period:
            params['period'] = period
            
        response = await self.http_client.get(
            '/analytics/referrers',
            params=params,
            config=config
        )
        
        return ReferrerStats(**response['data'])

    async def get_real_time_stats(
        self,
        config: Optional[RequestConfig] = None
    ) -> RealTimeStats:
        """Get real-time analytics statistics.
        
        Args:
            config: Request configuration
            
        Returns:
            RealTimeStats: Real-time statistics
        """
        response = await self.http_client.get('/analytics/realtime', config=config)
        return RealTimeStats(**response['data'])

    async def export_analytics(
        self,
        url_ids: Optional[List[int]] = None,
        format: str = 'csv',
        period: Optional[str] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        include_details: bool = False,
        filters: Optional[AnalyticsFilter] = None,
        config: Optional[RequestConfig] = None
    ) -> Dict[str, Any]:
        """Export analytics data.
        
        Args:
            url_ids: Specific URL IDs to export
            format: Export format (csv, json, xlsx)
            period: Time period (24h, 7d, 30d, etc.)
            start_date: Start date for export
            end_date: End date for export
            include_details: Include detailed click data
            filters: Additional filters
            config: Request configuration
            
        Returns:
            Dict[str, Any]: Export information
        """
        request = AnalyticsExportRequest(
            url_ids=url_ids,
            format=format,
            period=period,
            start_date=start_date,
            end_date=end_date,
            include_details=include_details,
            filters=filters
        )
        
        response = await self.http_client.post(
            '/analytics/export',
            data=request.dict(exclude_none=True),
            config=config
        )
        
        return response['data']

File: sdk/javascript/src/services/analytics.ts
import { HTTPClient } from '../http-client';
import { API_ENDPOINTS } from '../config';
import { 
  URLAnalytics,
  DashboardStats,
  URLPerformance,
  CountryStat,
  CityStat,
  DeviceStat,
  BrowserStat,
  ReferrerStat,
  Click,
  RequestConfig 
} from '../types';

export interface TimelineData {
  date: string;
  clicks: number;
  uniqueClicks: number;
}

export interface GeoStats {
  countries: CountryStat[];
  cities: CityStat[];
  totalClicks: number;
  uniqueCountries: number;
  uniqueCities: number;
}

export interface DeviceStats {
  devices: DeviceStat[];
  browsers: BrowserStat[];
  operatingSystems: DeviceStat[];
  totalClicks: number;
}

export interface ReferrerStats {
  referrers: ReferrerStat[];
  domains: ReferrerStat[];
  totalClicks: number;
  directClicks: number;
}

export type ExportFormat = 'csv' | 'json' | 'xlsx';
export type TimePeriod = '24h' | '7d' | '30d' | '90d' | '1y' | 'all';

export class AnalyticsService {
  constructor(private client: HTTPClient) {}

  /**
   * Get dashboard analytics overview
   */
  async getDashboard(config?: RequestConfig): Promise<DashboardStats> {
    return this.client.get<DashboardStats>(API_ENDPOINTS.ANALYTICS_DASHBOARD, config);
  }

  /**
   * Get global analytics statistics
   */
  async getGlobalStats(config?: RequestConfig): Promise<{
    totalURLs: number;
    totalClicks: number;
    totalUsers: number;
    clicksToday: number;
    urlsCreatedToday: number;
    averageClicksPerURL: number;
  }> {
    return this.client.get(API_ENDPOINTS.ANALYTICS_GLOBAL, config);
  }

  /**
   * Get top performing URLs
   */
  async getTopURLs(
    limit?: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<URLPerformance[]> {
    const params = new URLSearchParams();
    if (limit) params.set('limit', limit.toString());
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_TOP_URLS}${params.toString() ? `?${params.toString()}` : ''}`;
    const response = await this.client.get<{ urls: URLPerformance[] }>(url, config);
    return response.urls;
  }

  /**
   * Get detailed analytics for a specific URL
   */
  async getURLAnalytics(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<URLAnalytics> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL(urlId)}${params.toString() ? `?${params.toString()}` : ''}`;
    return this.client.get<URLAnalytics>(url, config);
  }

  /**
   * Get click timeline for a URL
   */
  async getClickTimeline(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<TimelineData[]> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL_TIMELINE(urlId)}${params.toString() ? `?${params.toString()}` : ''}`;
    const response = await this.client.get<{ timeline: TimelineData[] }>(url, config);
    return response.timeline;
  }

  /**
   * Get geographic statistics for a URL
   */
  async getGeographicStats(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<GeoStats> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL_GEO(urlId)}${params.toString() ? `?${params.toString()}` : ''}`;
    return this.client.get<GeoStats>(url, config);
  }

  /**
   * Get device and browser statistics for a URL
   */
  async getDeviceStats(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<DeviceStats> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL_DEVICES(urlId)}${params.toString() ? `?${params.toString()}` : ''}`;
    return this.client.get<DeviceStats>(url, config);
  }

  /**
   * Get referrer statistics for a URL
   */
  async getReferrerStats(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<ReferrerStats> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL_REFERRERS(urlId)}${params.toString() ? `?${params.toString()}` : ''}`;
    return this.client.get<ReferrerStats>(url, config);
  }

  /**
   * Export analytics data
   */
  async exportAnalytics(
    format: ExportFormat,
    options: {
      urlIds?: number[];
      period?: TimePeriod;
      startDate?: string;
      endDate?: string;
      includeDetails?: boolean;
    } = {},
    config?: RequestConfig
  ): Promise<Blob> {
    const params = new URLSearchParams();
    params.set('format', format);
    
    if (options.urlIds?.length) {
      params.set('urlIds', options.urlIds.join(','));
    }
    if (options.period) params.set('period', options.period);
    if (options.startDate) params.set('startDate', options.startDate);
    if (options.endDate) params.set('endDate', options.endDate);
    if (options.includeDetails) params.set('includeDetails', 'true');

    const url = `${API_ENDPOINTS.ANALYTICS_EXPORT}?${params.toString()}`;
    
    // Override request config to handle blob response
    const requestConfig = {
      ...config,
      headers: {
        ...config?.headers,
        'Accept': this.getAcceptHeader(format),
      },
    };

    const response = await this.client.get<ArrayBuffer>(url, requestConfig);
    return new Blob([response], { type: this.getMimeType(format) });
  }

  /**
   * Get real-time analytics (last 30 minutes)
   */
  async getRealTimeStats(config?: RequestConfig): Promise<{
    activeClicks: number;
    recentClicks: Click[];
    topActiveURLs: URLPerformance[];
    clicksPerMinute: { time: string; clicks: number }[];
  }> {
    return this.client.get(`${API_ENDPOINTS.ANALYTICS_DASHBOARD}/realtime`, config);
  }

  /**
   * Get analytics comparison between two periods
   */
  async getComparison(
    urlId: number,
    currentPeriod: TimePeriod,
    previousPeriod: TimePeriod,
    config?: RequestConfig
  ): Promise<{
    current: URLAnalytics;
    previous: URLAnalytics;
    growth: {
      clicks: number;
      uniqueClicks: number;
      clicksPercentage: number;
      uniqueClicksPercentage: number;
    };
  }> {
    const params = new URLSearchParams();
    params.set('current', currentPeriod);
    params.set('previous', previousPeriod);

    const url = `${API_ENDPOINTS.ANALYTICS_URL(urlId)}/compare?${params.toString()}`;
    return this.client.get(url, config);
  }

  /**
   * Get analytics for multiple URLs
   */
  async getBulkAnalytics(
    urlIds: number[],
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<URLAnalytics[]> {
    const params = new URLSearchParams();
    params.set('urlIds', urlIds.join(','));
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_DASHBOARD}/bulk?${params.toString()}`;
    const response = await this.client.get<{ analytics: URLAnalytics[] }>(url, config);
    return response.analytics;
  }

  /**
   * Get click heatmap data for time-based visualization
   */
  async getClickHeatmap(
    urlId: number,
    period?: TimePeriod,
    config?: RequestConfig
  ): Promise<{
    hourly: { hour: number; clicks: number }[];
    daily: { day: number; clicks: number }[];
    weekly: { week: number; clicks: number }[];
  }> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.ANALYTICS_URL(urlId)}/heatmap${params.toString() ? `?${params.toString()}` : ''}`;
    return this.client.get(url, config);
  }

  private getAcceptHeader(format: ExportFormat): string {
    switch (format) {
      case 'csv':
        return 'text/csv';
      case 'json':
        return 'application/json';
      case 'xlsx':
        return 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
      default:
        return 'application/octet-stream';
    }
  }

  private getMimeType(format: ExportFormat): string {
    switch (format) {
      case 'csv':
        return 'text/csv';
      case 'json':
        return 'application/json';
      case 'xlsx':
        return 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
      default:
        return 'application/octet-stream';
    }
  }
}